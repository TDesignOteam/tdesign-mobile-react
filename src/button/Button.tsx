import React, { forwardRef } from 'react';
import classnames from 'classnames';
import { LoadingIcon } from 'tdesign-icons-react';
import useConfig from '../_util/useConfig';
import parseTNode from '../_util/parseTNode';
import { TdButtonProps } from './type';
import noop from '../_util/noop';
import { buttonDefaultProps } from './defaultProps';

export interface ButtonProps
  extends TdButtonProps,
    Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'content' | 'children'> {}

const Button = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const {
    className = '',
    style,
    block,
    children,
    content = '',
    disabled,
    ghost,
    icon = null,
    suffix = null,
    loading,
    shape,
    size,
    theme,
    type,
    variant,
    onClick = noop,
    loadingProps = {},
  } = props;
  const { classPrefix } = useConfig();

  const childNode = content || children;

  return (
    <button
      ref={ref}
      type={type}
      className={classnames(
        [
          `${classPrefix}-button`,
          `${classPrefix}-button--size-${size}`,
          `${classPrefix}-button--${variant}`,
          `${classPrefix}-button--${theme}`,
          `${classPrefix}-button--${shape}`,
          className,
        ],
        {
          [`${classPrefix}-button--ghost`]: ghost,
          [`${classPrefix}-button--loading`]: loading,
          [`${classPrefix}-button--disabled`]: disabled,
          [`${classPrefix}-button--block`]: block,
        },
      )}
      style={style}
      onClick={!loading && !disabled ? onClick : undefined}
      disabled={disabled || loading}
    >
      {loading ? <LoadingIcon {...loadingProps} /> : parseTNode(icon)}
      {childNode && <span className={`${classPrefix}-button__content`}> {parseTNode(childNode)}</span>}
      {parseTNode(suffix)}
    </button>
  );
});

Button.displayName = 'Button';
Button.defaultProps = buttonDefaultProps;

export default Button;
