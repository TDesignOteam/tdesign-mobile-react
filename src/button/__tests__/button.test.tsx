import React from 'react';
import { render } from '@test/utils';
import Button from '../Button';

describe('Button 组件测试', () => {
  const ButtonText = '按钮组件';
  test('content', async () => {
    const { queryByText } = render(<Button content={ButtonText} />);
    expect(queryByText(ButtonText)).toBeInTheDocument();
  });
});
