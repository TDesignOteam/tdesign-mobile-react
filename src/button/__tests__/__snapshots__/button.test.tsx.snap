// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button 组件测试 > content 1`] = `
<span
  class="t-button__content"
>
   
  按钮组件
</span>
`;
